packages/flutter_tools/lib/src/application_package.dart

```diff
@@ -349,34 +360,127 @@ class ApplicationPackageStore {
+        orElse: () => null);
+  }
+
+  Iterable<Entry> allElements(String name) {
+    return children.where(
+            (Entry e) => e is Element && e.name.startsWith(name));
+  }
+}
+
+class Attribute extends Entry {
+  String key;
+  String value;
+
+  Attribute.fromLine(String line, Element parent) {
+    //     A: android:label(0x01010001)="hello_world" (Raw: "hello_world")
+    final List<String> keyVal = line.substring(line.indexOf('A: ') + 3).split('=');
+    key = keyVal[0];
+    value = keyVal[1];
+    level = line.length - line.trimLeft().length;
+    this.parent = parent;
+  }
+}
+
 class ApkManifestData {
   ApkManifestData._(this._data);
 
-  static ApkManifestData parseFromAaptBadging(String data) {
+  static ApkManifestData parseFromXmlDump(String data) {
     if (data == null || data.trim().isEmpty)
       return null;
 
-    // package: name='io.flutter.gallery' versionCode='1' versionName='0.0.1' platformBuildVersionName='NMR1'
-    // launchable-activity: name='io.flutter.app.FlutterActivity'  label='' icon=''
-    final Map<String, Map<String, String>> map = <String, Map<String, String>>{};
+    final List<String> lines = data.split('\n');
+    assert(lines.length > 3);
 
-    final RegExp keyValueRegex = new RegExp(r"(\S+?)='(.*?)'");
+    final Element manifest = new Element.fromLine(lines[1], null);
+    Element currentElement = manifest;
 
-    for (String line in data.split('\n')) {
-      final int index = line.indexOf(':');
-      if (index != -1) {
-        final String name = line.substring(0, index);
-        line = line.substring(index + 1).trim();
+    for (String line in lines.skip(2)) {
+      final String trimLine = line.trimLeft();
+      final int level = line.length - trimLine.length;
 
-        final Map<String, String> entries = <String, String>{};
-        map[name] = entries;
+      // Handle level out
+      while(level <= currentElement.level) {
+        currentElement = currentElement.parent;
+      }
 
-        for (Match m in keyValueRegex.allMatches(line)) {
-          entries[m.group(1)] = m.group(2);
+      if (level > currentElement.level) {
+        switch (trimLine[0]) {
+          case 'A':
+            currentElement
+                .addChild(new Attribute.fromLine(line, currentElement));
+            break;
+          case 'E':
+            final Element element = new Element.fromLine(line,currentElement);
+            currentElement.addChild(element);
+            currentElement = element;
         }
       }
     }
 
+    final Element application = manifest.firstElement('application');
+    assert(application != null);
+
+    final Iterable<Entry> activities = application.allElements('activity');
+
+    Element launchActivity;
+    for (Element activity in activities) {
+      final Attribute enabled = activity.firstAttribute('android:enabled');
+      if (enabled == null || enabled.value.contains('0xffffffff')) {
+        launchActivity = activity;
```

Comment:
This appears to imply that we get the last launchable activity. Android Studio seems to start the first such thing.